sdlib/d/gc/slab.d

```diff
@@ -23,6 +25,159 @@ struct SlabAllocGeometry {
 	}
 }
 
+struct SlabAllocInfo {
+public:
+	bool hasMetaData = false;
+
+	this(PageDescriptor pd, const void* ptr) {
+		assert(pd.isSlab(), "Expected a slab!");
+
+		e = pd.extent;
+		sg = SlabAllocGeometry(pd, ptr);
+		allowsMetaData = sizeClassSupportsMetadata(pd.sizeClass);
+		hasMetaData = allowsMetaData && e.hasMetaData(sg.index);
+	}
+
+	@property
+	auto address() {
+		return sg.address;
+	}
+
+	@property
+	size_t slotCapacity() {
+		return sg.size;
+	}
+
+	@property
+	size_t usedCapacity() {
+		return allowsMetaData ? sg.size - freeSpace : 0;
+	}
+
+	bool setUsedCapacity(size_t size) {
+		if (!allowsMetaData || (size > slotCapacity)) {
+			return false;
+		}
+
+		setFreeSpace(sg.size - size);
+		return true;
+	}
+
+private:
+	SlabAllocGeometry sg;
+	Extent* e;
+	bool allowsMetaData = false;
+
+	void setMetaData(bool value) {
```

Comment:
This is always called with a constant, it doesn't seems like it's useful in this form.